src/years.rs

```diff
@@ -0,0 +1,115 @@
+// I usually develop with this clippy flags
+// but I can remove once the this module is done
+#![warn(clippy::all,
+/*#![warn(*/clippy::pedantic,
+		clippy::perf,
+        clippy::nursery,
+        // clippy::cargo,
+        clippy::unwrap_used,
+        clippy::expect_used)]
+
+use num_integer::Integer;
+
+use super::{datetime::DateTime, NaiveDate, NaiveDateTime};
+use crate::traits::Datelike;
+use crate::TimeZone;
+
+struct Years(u32); // I believe the fields of the struct should be private
+
+impl Years {
+    fn new(num_years: u32) -> Years {
+        Self(num_years)
+    }
+}
+
+// would to the same for all the others
+impl<Tz: TimeZone> DateTime<Tz> {
+    pub fn checked_add_years(self, num_years: Years) -> Option<DateTime<Tz>> {}
+
+    pub fn checked_sub_years(self, num_years: Years) -> Option<DateTime<Tz>> {}
+}
+
+impl NaiveDateTime {
+    pub fn checked_add_years(self, num_years: Years) -> Option<NaiveDateTime> {}
+
+    pub fn checked_sub_years(self, num_years: Years) -> Option<NaiveDateTime> {}
+}
+
+// think about a better logic to use `checked_add`
+// currently trying to finish a simple implementation
+// is it ok to assum it'll only receive positive numbers? I think so
+fn get_current_or_next_leap_year(year: u32) -> u32 {
+    if !year.is_even() {
+        year += 1;
+    }
+    while !is_leap_year(year) {
+        year += 1;
+    }
+    year
+}
+
+// almost surely there is a more precise way to implement this
+fn is_leap_year(year: i32) -> bool {
```

Comment:
you can use ordinal flags and then use the ndays method to figure out whether the target year is a leap year or not. You can construct the ordinal flags directly from the target year itself

Reply:
Thank you for the pointers! I'll surely make the requested changes! This one in special will be of great help! really appreciated!

I'll move the other impl blocks to the proper locations too!